src/chains/filecoin/filecoin/src/database.ts

```diff
@@ -1,7 +1,173 @@
-export type DatabaseOptions = {};
+import { AbstractLevelDOWN } from "abstract-leveldown";
+import Emittery from "emittery";
+import { dir, setGracefulCleanup } from "tmp-promise";
+import levelup, { LevelUp } from "levelup";
+import { FilecoinInternalOptions } from "@ganache/filecoin-options";
+const leveldown = require("leveldown");
+const sub = require("subleveldown");
+const encode = require("encoding-down");
 
-export default class Database implements DatabaseOptions {
-  constructor(options: DatabaseOptions) {
-    Object.assign(this, options);
+setGracefulCleanup();
+const tmpOptions = { prefix: "ganache-core_", unsafeCleanup: true };
+const noop = () => Promise.resolve();
+
+export default class Database extends Emittery {
+  readonly #options: FilecoinInternalOptions["database"];
+  #cleanupDirectory = noop;
+  #closed = false;
+
+  public directory: string | null = null;
+  public db: LevelUp | null = null;
+
+  public tipsets: LevelUp | null = null;
+  public blocks: LevelUp | null = null;
+
+  #initialized: boolean = false;
+  get initialized() {
+    return this.#initialized;
+  }
+
+  #rootStore: AbstractLevelDOWN | null = null;
+
+  /**
+   * The Database handles the creation of the database, and all access to it.
+   * Once the database has been fully initialized it will emit a `ready`
+   * event.
+   * @param options Supports one of two options: `db` (a leveldown compliant
+   * store instance) or `dbPath` (the path to store/read the db instance)
+   * @param blockchain
+   */
+  constructor(options: FilecoinInternalOptions["database"]) {
+    super();
+
+    this.#options = options;
+    this.#initialize();
   }
+
+  #initialize = async () => {
+    const levelupOptions: any = { valueEncoding: "binary" };
+    const store = this.#options.db;
+    let db: LevelUp;
+    if (store) {
+      this.#rootStore = encode(store, levelupOptions);
+      // @ts-ignore - I get an error in ts-node that I don't get running ttsc:
+      // Argument of type 'AbstractLevelDOWN<any, any>' is not assignable to parameter of type 'string'
+      // But this error just doesn't make any sense. I'm just going to ignore the error
```

Comment:
I just started getting this error after updating some dependencies. Some type def in some transitive package must have erroneously changed things on us :expressionless:

Reply:
Did you want me to get to the bottom of it or not bother?

Reply:
eh, not bother.